WD-Firmware/comrv/comrv.c

```diff
@@ -18,34 +18,765 @@
+         /* save the selected multi group entry */
+         usSelectedMultiGroupEntry = unToken.stFields.overlayGroupID;
+      }
+#endif /* D_COMRV_MULTI_GROUP_SUPPORT */
+
+      /* we need to make sure that from this point
+         we won't have new overlay requests */
+      M_COMRV_ENTER_CRITICAL_SECTION();
+
+      /* get eviction candidates according to the requested pOverlayGroupSize */
+      ucNumOfEvictionCandidates = comrvGetEvictionCandidates(usOverlayGroupSize, ucEvictCandidateList);
+
+      ucEntryIndex = 0;
+      /* we need to handle heap fragmentation since we got more
+         than 1 eviction candidate */
+      if (ucNumOfEvictionCandidates > 1)
+      {
+         /* no handling to the last eviction candidate */
+         ucNumOfEvictionCandidates--;
+         while (ucEntryIndex < ucNumOfEvictionCandidates)
+         {
+            /* get the candidate entry */
+            pEntry = &stComrvCB.stOverlayCache[ucEvictCandidateList[ucEntryIndex]];
+            /* calc the source address */
+            pAddress = ((u08_t*)pEntry->pFixedEntryAddress + M_COMRV_GET_OVL_GROUP_SIZE_IN_BYTES(pEntry->unToken));
+            /* perform code copy */
+            comrvMemcpyHook(pEntry->pFixedEntryAddress, pAddress,
+                  stComrvCB.stOverlayCache[ucEvictCandidateList[ucEntryIndex]].pFixedEntryAddress - pAddress);
+
+            /* after code copy we need to align the entries structures */
+            for(ucIndex = ucEvictCandidateList[ucEntryIndex] ; ucIndex < ucEvictCandidateList[ucEntryIndex+1] ; ucIndex++)
+            {
+               pEntry = &stComrvCB.stOverlayCache[ucIndex + stComrvCB.stOverlayCache[ucIndex].unProperties.stFields.ucSizeInMinGroupSizeUnits];
+#ifdef D_COMRV_OVL_DATA_SUPPORT
+               /* an overlay data is present when handling defragmentation */
+               if (pEntry->unProperties.ucData)
+               {
+                  M_COMRV_EXIT_CRITICAL_SECTION();
+                  comrvNotificationHook(D_COMRV_OVL_DATA_DEFRAG_ERR, unToken.ucValue);
+               }
+#endif /* D_COMRV_OVL_DATA_SUPPORT */
+               stComrvCB.stOverlayCache[ucIndex].unProperties = pEntry->unProperties;
+               stComrvCB.stOverlayCache[ucIndex].unToken.uiValue = pEntry->unToken.uiValue;
+#ifdef D_COMRV_EVICTION_LRU
+               stComrvCB.stOverlayCache[pEntry->ucPrevIndex].ucNextIndex = ucIndex;
+               stComrvCB.stOverlayCache[pEntry->ucNextIndex].ucPrevIndex = ucIndex;
+#elif defined(D_COMRV_EVICTION_LFU)
+#elif defined(D_COMRV_EVICTION_MIX_LRU_LFU)
+#endif /* D_COMRV_EVICTION_LRU */
+            }
+            ucEntryIndex++;
+         }
+         ucNumOfEvictionCandidates++;
+      }
+      /* this means the end user locked all entries */
+      else if (ucNumOfEvictionCandidates == 0)
+      {
+         M_COMRV_EXIT_CRITICAL_SECTION();
+         comrvNotificationHook(D_COMRV_NO_AVAILABLE_ENTRY_ERR, unToken.uiValue);
```

Comment:
implementation of error_hook
implementation of notification_hook

Reply:
done in d17a18a